2019/05/5.rb

```diff
@@ -0,0 +1,169 @@
+
+      first_input_value = get_first_parameter_value(command, input, i)
+      second_input_value = get_second_parameter_value(command, input, i)
+      output_index = get_third_parameter_value(command, input, i)
+
+      result = first_input_value * second_input_value
+      input[output_index] = result.to_s
+
+    elsif opcode == :read
+
+      first_input_value = Integer(input[i+1])
+      input[first_input_value] = 5 # first input value, this is a dirty hack
+      
+    elsif opcode == :write
+    
+      first_input_index = Integer(input[i+1])
+      puts input[first_input_index]
+
+    elsif opcode == :jump_if_true
+      first_input_value = get_first_parameter_value(command, input, i)
+      second_input_value = get_second_parameter_value(command, input, i)
+
+      if first_input_value != 0
+        i = Integer(second_input_value)
+        next
+      end
+
+    elsif opcode == :jump_if_false
+      first_input_value = get_first_parameter_value(command, input, i)
+      second_input_value = get_second_parameter_value(command, input, i)
+
+      if first_input_value == 0
+        i = Integer(second_input_value)
+        next
+      end
+
+    elsif opcode == :less_than
+      first_input_value = get_first_parameter_value(command, input, i)
+      second_input_value = get_second_parameter_value(command, input, i)
+      output_index = Integer(input[i+3])
+      
+      if first_input_value < second_input_value
+        input[output_index] = "1"
+      else
+        input[output_index] = "0"
+      end
+
+    elsif opcode == :equals
+      first_input_value = get_first_parameter_value(command, input, i)
+      second_input_value = get_second_parameter_value(command, input, i)
+      output_index = Integer(input[i+3])
+      
+      parameters_equal = first_input_value == second_input_value
+      if parameters_equal
+      	input[output_index] = "1"
+      else
+        input[output_index] = "0"
+      end
+    else
+      puts "Weird opcode (#{opcode} at #{i}), command: #{command}, aborting..."
+    end
+    
+    number_of_parameters = number_of_parameters_for_opcode(opcode)
+    
+    i += number_of_parameters
+    i += 1 # next command
+  end
+end
+
+def get_opcode_for_command(command)
+  if /1$/ =~ command
+    return :add
+  elsif /2$/ =~ command
+    return :multiply
+  elsif /3$/ =~ command
+    return :read
+  elsif /4$/ =~ command
+    return :write
+  elsif /5$/ =~ command
+  	return :jump_if_true
+  elsif /6$/ =~ command
+    return :jump_if_false
+  elsif /7$/ =~ command
+    return :less_than
+  elsif /8$/ =~ command
+    return :equals
+  elsif /^99$/ =~ command
+    return :halt
+  end
+end
+
+def number_of_parameters_for_opcode(opcode)
+  if opcode == :halt
+    return 0
+  elsif opcode == :add || opcode == :multiply || opcode == :less_than || opcode == :equals
+    return 3
+  elsif opcode == :jump_if_false || opcode == :jump_if_true
+    return 2
+  elsif opcode == :read || opcode == :write
+    return 1
+  end
+end
```

Comment:
I know I said the thing about unnecessar returns already, but here it is also not necessary

```ruby
def number_of_parameters_for_opcode(opcode)
  if opcode == :halt
    0
  elsif opcode == :add || opcode == :multiply || opcode == :less_than || opcode == :equals
    3
  elsif opcode == :jump_if_false || opcode == :jump_if_true
    2
  elsif opcode == :read || opcode == :write
    1
  end
end
```